src/Reader/Common/Creator/ReaderFactory.php

```diff
@@ -27,11 +27,31 @@ public static function createFromFile(string $path): ReaderInterface
     {
         $extension = strtolower(pathinfo($path, PATHINFO_EXTENSION));
 
-        return match ($extension) {
+        $writer = match($extension) {
             'csv' => new CSVReader(),
             'xlsx' => new XLSXReader(),
             'ods' => new ODSReader(),
-            default => throw new UnsupportedTypeException('No readers supporting the given type: '.$extension),
+            default => null
         };
+
+        if ($writer) {
+            return $writer;
+        }
+
+        $mime_type = mime_content_type($path);
```

Comment:
I'd say this deserves a dedicated factory method

Reply:
Something like `createFromFileByMimeType`?

Reply:
:+1:

Reply:
I have committed and pushed the change to my branch, but I don't know if that propagates properly to the pull request.

I am new to contributing to GitHub projects so please forgive me if I happened to have done the wrong thing.

Reply:
If you have php installed localy, type `make` in the project to run a local build.

I'll take care of the `Lint composer.json` step, but please fix Coding Standards and Static Analysis.

Moreover, please add tests for the new API

Reply:
Will do. Should I also add an ignore line to the phpstan config as is done for `createFromFile`?

Reply:
I have added a file_exists() check to the `createFromFileByMimeType` method.

Tests have been added. I mirrored the tests from `ReaderFactoryTest` in `ReaderEntityFactoryTest`, it seems redundant to me, but I may be wrong so I added the same tests to it as well.

Reply:
Can you fix the failing steps of the build please?

Reply:
I feel like you are not building locally, and instead you're waiting for CI feedback, aren't you?

That's a waste of resources, please build locally by using `make` command and have it all green before pushing.

Reply:
> So it's not due to any code I introduced.

Right, just adapt to it and remove the ignored error pattern.

But you did you run the build locally and the `Coding Standards` build still fails in CI?

It's the first command in the `make` list, it should have been fixed by now

Reply:
I did run that, but then it wanted to change a bunch of other files as well, ones I haven't touched. I didn't feel like I should have pushed that. I could of course do that if that's preferred.

Here's what csfix changed after running:

```diff
diff --git a/benchmarks/OdsWriterBench.php b/benchmarks/OdsWriterBench.php
index 6727388..cab395f 100644
--- a/benchmarks/OdsWriterBench.php
+++ b/benchmarks/OdsWriterBench.php
@@ -48,7 +48,7 @@ final class OdsWriterBench
         $preg_match_all = preg_match_all('/<text:p>ods--(\d+)-\d<\/text:p>/', $endingContentXmlContents, $matches);
         if (false !== $preg_match_all && 0 < $preg_match_all) {
             $lastMatch = array_pop($matches);
-            $numWrittenRows = (int) (array_pop($lastMatch));
+            $numWrittenRows = (int) array_pop($lastMatch);
         }
 
         return $numWrittenRows;
diff --git a/benchmarks/XlsxWriterTrait.php b/benchmarks/XlsxWriterTrait.php
index 4c802c2..58ce725 100644
--- a/benchmarks/XlsxWriterTrait.php
+++ b/benchmarks/XlsxWriterTrait.php
@@ -39,7 +39,7 @@ trait XlsxWriterTrait
         $preg_match_all = preg_match_all('/<t.*>xlsx--(\d+)-\d+<\/t>/', $lastCharactersOfFile, $matches);
         if (false !== $preg_match_all && 0 < $preg_match_all) {
             $lastMatch = array_pop($matches);
-            $lastRowNumber = (int) (array_pop($lastMatch));
+            $lastRowNumber = (int) array_pop($lastMatch);
         }
 
         return $lastRowNumber;
diff --git a/src/Common/Helper/FileSystemHelper.php b/src/Common/Helper/FileSystemHelper.php
index a57c11c..e3940d0 100644
--- a/src/Common/Helper/FileSystemHelper.php
+++ b/src/Common/Helper/FileSystemHelper.php
@@ -156,7 +156,7 @@ final class FileSystemHelper implements FileSystemHelperInterface
         if (false === $operationFolderRealPath) {
             throw new IOException("Folder not found: {$operationFolderRealPath}");
         }
-        $isInBaseFolder = (str_starts_with($operationFolderRealPath, $this->baseFolderRealPath));
+        $isInBaseFolder = str_starts_with($operationFolderRealPath, $this->baseFolderRealPath);
         if (!$isInBaseFolder) {
             throw new IOException("Cannot perform I/O operation outside of the base folder: {$this->baseFolderRealPath}");
         }
diff --git a/src/Reader/Common/Creator/ReaderFactory.php b/src/Reader/Common/Creator/ReaderFactory.php
index 2603e7b..c72bcfa 100644
--- a/src/Reader/Common/Creator/ReaderFactory.php
+++ b/src/Reader/Common/Creator/ReaderFactory.php
@@ -4,8 +4,8 @@ declare(strict_types=1);
 
 namespace OpenSpout\Reader\Common\Creator;
 
-use OpenSpout\Common\Exception\UnsupportedTypeException;
 use OpenSpout\Common\Exception\IOException;
+use OpenSpout\Common\Exception\UnsupportedTypeException;
 use OpenSpout\Reader\CSV\Reader as CSVReader;
 use OpenSpout\Reader\ODS\Reader as ODSReader;
 use OpenSpout\Reader\ReaderInterface;
diff --git a/src/Reader/ODS/Helper/CellValueFormatter.php b/src/Reader/ODS/Helper/CellValueFormatter.php
index 645135a..2a4c7e7 100644
--- a/src/Reader/ODS/Helper/CellValueFormatter.php
+++ b/src/Reader/ODS/Helper/CellValueFormatter.php
@@ -142,7 +142,7 @@ final class CellValueFormatter
      */
     private function formatBooleanCellValue(DOMElement $node): bool
     {
-        return (bool) ($node->getAttribute(self::XML_ATTRIBUTE_BOOLEAN_VALUE));
+        return (bool) $node->getAttribute(self::XML_ATTRIBUTE_BOOLEAN_VALUE);
     }
 
     /**
diff --git a/src/Reader/Wrapper/XMLReader.php b/src/Reader/Wrapper/XMLReader.php
index 0773f16..639dd9f 100644
--- a/src/Reader/Wrapper/XMLReader.php
+++ b/src/Reader/Wrapper/XMLReader.php
@@ -172,7 +172,7 @@ final class XMLReader extends \XMLReader
          *
          * @see https://github.com/box/spout/issues/233
          */
-        $hasPrefix = (str_contains($nodeName, ':'));
+        $hasPrefix = str_contains($nodeName, ':');
         $currentNodeName = ($hasPrefix) ? $this->name : $this->localName;
 
         return $this->nodeType === $nodeType && $currentNodeName === $nodeName;
diff --git a/src/Reader/XLSX/Helper/CellHelper.php b/src/Reader/XLSX/Helper/CellHelper.php
index ccfdc54..19c4c24 100644
--- a/src/Reader/XLSX/Helper/CellHelper.php
+++ b/src/Reader/XLSX/Helper/CellHelper.php
@@ -48,7 +48,7 @@ final class CellHelper
         // Also, not using the pow() function because it's slooooow...
         switch ($columnLength) {
             case 1:
-                $columnIndex = (self::columnLetterToIndexMapping[$columnLetters]);
+                $columnIndex = self::columnLetterToIndexMapping[$columnLetters];
 
                 break;
 
diff --git a/src/Reader/XLSX/Manager/SharedStringsCaching/MemoryLimit.php b/src/Reader/XLSX/Manager/SharedStringsCaching/MemoryLimit.php
index c660ee4..e9d7628 100644
--- a/src/Reader/XLSX/Manager/SharedStringsCaching/MemoryLimit.php
+++ b/src/Reader/XLSX/Manager/SharedStringsCaching/MemoryLimit.php
@@ -29,7 +29,7 @@ final class MemoryLimit
         }
 
         if (1 === preg_match('/(\d+)([bkmgt])b?/', $memoryLimitFormatted, $matches)) {
-            $amount = (int) ($matches[1]);
+            $amount = (int) $matches[1];
             $unit = $matches[2];
 
             switch ($unit) {
diff --git a/src/Reader/XLSX/Manager/StyleManager.php b/src/Reader/XLSX/Manager/StyleManager.php
index 086af33..aa6bb01 100644
--- a/src/Reader/XLSX/Manager/StyleManager.php
+++ b/src/Reader/XLSX/Manager/StyleManager.php
@@ -170,7 +170,7 @@ class StyleManager implements StyleManagerInterface
     {
         while ($xmlReader->read()) {
             if ($xmlReader->isPositionedOnStartingNode(self::XML_NODE_NUM_FMT)) {
-                $numFmtId = (int) ($xmlReader->getAttribute(self::XML_ATTRIBUTE_NUM_FMT_ID));
+                $numFmtId = (int) $xmlReader->getAttribute(self::XML_ATTRIBUTE_NUM_FMT_ID);
                 $formatCode = $xmlReader->getAttribute(self::XML_ATTRIBUTE_FORMAT_CODE);
                 \assert(null !== $formatCode);
                 $this->customNumberFormats[$numFmtId] = $formatCode;
diff --git a/src/Reader/XLSX/Manager/WorkbookRelationshipsManager.php b/src/Reader/XLSX/Manager/WorkbookRelationshipsManager.php
index 83fd0c5..e8c99ef 100644
--- a/src/Reader/XLSX/Manager/WorkbookRelationshipsManager.php
+++ b/src/Reader/XLSX/Manager/WorkbookRelationshipsManager.php
@@ -58,7 +58,7 @@ final class WorkbookRelationshipsManager
             ?? $workbookRelationships[self::RELATIONSHIP_TYPE_SHARED_STRINGS_STRICT];
 
         // the file path can be relative (e.g. "styles.xml") or absolute (e.g. "/xl/styles.xml")
-        $doesContainBasePath = (str_contains($sharedStringsXMLFilePath, self::BASE_PATH));
+        $doesContainBasePath = str_contains($sharedStringsXMLFilePath, self::BASE_PATH);
         if (!$doesContainBasePath) {
             // make sure we return an absolute file path
             $sharedStringsXMLFilePath = self::BASE_PATH.$sharedStringsXMLFilePath;
@@ -99,7 +99,7 @@ final class WorkbookRelationshipsManager
             ?? $workbookRelationships[self::RELATIONSHIP_TYPE_STYLES_STRICT];
 
         // the file path can be relative (e.g. "styles.xml") or absolute (e.g. "/xl/styles.xml")
-        $doesContainBasePath = (str_contains($stylesXMLFilePath, self::BASE_PATH));
+        $doesContainBasePath = str_contains($stylesXMLFilePath, self::BASE_PATH);
         if (!$doesContainBasePath) {
             // make sure we return a full path
             $stylesXMLFilePath = self::BASE_PATH.$stylesXMLFilePath;
diff --git a/src/Writer/XLSX/Manager/WorksheetManager.php b/src/Writer/XLSX/Manager/WorksheetManager.php
index f219527..4ad96c1 100644
--- a/src/Writer/XLSX/Manager/WorksheetManager.php
+++ b/src/Writer/XLSX/Manager/WorksheetManager.php
@@ -193,7 +193,7 @@ final class WorksheetManager implements WorksheetManagerInterface
         if ($cell instanceof Cell\StringCell) {
             $cellXML .= $this->getCellXMLFragmentForNonEmptyString($cell->getValue());
         } elseif ($cell instanceof Cell\BooleanCell) {
-            $cellXML .= ' t="b"><v>'.(int) ($cell->getValue()).'</v></c>';
+            $cellXML .= ' t="b"><v>'.(int) $cell->getValue().'</v></c>';
         } elseif ($cell instanceof Cell\NumericCell) {
             $cellXML .= '><v>'.$cell->getValue().'</v></c>';
         } elseif ($cell instanceof Cell\FormulaCell) {
diff --git a/tests/Reader/Common/Creator/ReaderEntityFactoryTest.php b/tests/Reader/Common/Creator/ReaderEntityFactoryTest.php
index 47e382a..9eaa590 100644
--- a/tests/Reader/Common/Creator/ReaderEntityFactoryTest.php
+++ b/tests/Reader/Common/Creator/ReaderEntityFactoryTest.php
@@ -4,8 +4,8 @@ declare(strict_types=1);
 
 namespace OpenSpout\Reader\Common\Creator;
 
-use OpenSpout\Common\Exception\UnsupportedTypeException;
 use OpenSpout\Common\Exception\IOException;
+use OpenSpout\Common\Exception\UnsupportedTypeException;
 use OpenSpout\TestUsingResource;
 use PHPUnit\Framework\TestCase;
 
diff --git a/tests/Reader/Common/Creator/ReaderFactoryTest.php b/tests/Reader/Common/Creator/ReaderFactoryTest.php
index 67f644f..53a2db6 100644
--- a/tests/Reader/Common/Creator/ReaderFactoryTest.php
+++ b/tests/Reader/Common/Creator/ReaderFactoryTest.php
@@ -4,8 +4,8 @@ declare(strict_types=1);
 
 namespace OpenSpout\Reader\Common\Creator;
 
-use OpenSpout\Common\Exception\UnsupportedTypeException;
 use OpenSpout\Common\Exception\IOException;
+use OpenSpout\Common\Exception\UnsupportedTypeException;
 use OpenSpout\TestUsingResource;
 use PHPUnit\Framework\TestCase;
 
diff --git a/tests/Writer/ODS/WriterTest.php b/tests/Writer/ODS/WriterTest.php
index 6ce77e6..2863b41 100644
--- a/tests/Writer/ODS/WriterTest.php
+++ b/tests/Writer/ODS/WriterTest.php
@@ -287,7 +287,7 @@ final class WriterTest extends TestCase
 
         if (1 === $expectedNumTableCells) {
             $tableCellNode = $tableCellNodes->item(0);
-            $numColumnsRepeated = (int) ($tableCellNode->getAttribute('table:number-columns-repeated'));
+            $numColumnsRepeated = (int) $tableCellNode->getAttribute('table:number-columns-repeated');
             self::assertSame($expectedNumColumnsRepeated, $numColumnsRepeated);
         } else {
             foreach ($tableCellNodes as $tableCellNode) {
diff --git a/tests/Writer/ODS/WriterWithStyleTest.php b/tests/Writer/ODS/WriterWithStyleTest.php
index fdfc40a..e2adb0b 100644
--- a/tests/Writer/ODS/WriterWithStyleTest.php
+++ b/tests/Writer/ODS/WriterWithStyleTest.php
@@ -295,7 +295,7 @@ final class WriterWithStyleTest extends TestCase
 
         $styles = [
             (new Style())->setBorder($borderBottomGreenThickSolid),
-            (new Style()),
+            new Style(),
             (new Style())->setBorder($borderTopRedThinDashed),
             (new Style())->setBorder($borderLeft),
         ];
diff --git a/tests/Writer/XLSX/WriterWithStyleTest.php b/tests/Writer/XLSX/WriterWithStyleTest.php
index 7544aea..ae8d32b 100644
--- a/tests/Writer/XLSX/WriterWithStyleTest.php
+++ b/tests/Writer/XLSX/WriterWithStyleTest.php
@@ -432,7 +432,7 @@ final class WriterWithStyleTest extends TestCase
 
         $styles = [
             (new Style())->setBorder($borderBottomGreenThickSolid),
-            (new Style()),
+            new Style(),
             (new Style())->setBorder($borderTopRedThinDashed),
         ];
```

Reply:
Generally speaking, a green build is always needed for merge in any case.

Please try to get to it :muscle:

Reply:
Alright, everything ran well on my end without any errors. So that would mean all checks should now come up green.

Reply:
The CI test `Tests (windows-latest, 8.1, none)` seems to fail because it does not have the `mime_content_type()` function. So I added ext-fileinfo to the composer.json requirements.

Maybe it would be a good idea to add the composer-require-checker to the make file as well.

Reply:
The composer lint now succeeds, but the windows unit tests fail because it can't seem to install the added dependency. Here I'm stuck as I don't know how to resolve this particular issue.